dipy/io/stateful_tractogram.py

```diff
@@ -0,0 +1,746 @@
+                    self._voxel_sizes)
+                self._tractogram.apply_affine(self._affine)
+                self._space = Space.RASMM
+                logging.info('Moved streamlines from voxmm to rasmm')
+        else:
+            logging.warning('Wrong initial space for this function')
+            return
+
+    def _rasmm_to_voxmm(self):
+        """ Unsafe function to transform streamlines """
+        if self._space == Space.RASMM:
+            if self._tractogram.streamlines.data.size > 0:
+                self._tractogram.apply_affine(self._inv_affine)
+                self._tractogram.streamlines._data *= np.asarray(
+                    self._voxel_sizes)
+                self._space = Space.VOXMM
+                logging.info('Moved streamlines from rasmm to voxmm')
+        else:
+            logging.warning('Wrong initial space for this function')
+            return
+
+    def _shift_voxel_origin(self):
+        """ Unsafe function to switch the origin from center to corner
+        and vice versa """
+        shift = np.asarray([0.5, 0.5, 0.5])
+        if self._space == Space.VOXMM:
+            shift = shift * self._voxel_sizes
+        elif self._space == Space.RASMM:
+            tmp_affine = np.eye(4)
+            tmp_affine[0:3, 0:3] = self._affine[0:3, 0:3]
+            shift = apply_affine(tmp_affine, shift)
+        if self._shifted_origin:
+            shift *= -1
+
+        self._tractogram.streamlines._data += shift
+        if not self._shifted_origin:
+            logging.info('Origin moved to the center of voxel')
+        else:
+            logging.info('Origin moved to the corner of voxel')
+
+        self._shifted_origin = not self._shifted_origin
+
+
+def save_tractogram(sft, filename, bbox_valid_check=True):
+    """ Save the stateful tractogram in any format (trk, tck, fib, dpy)
+
+    Parameters
+    ----------
+    sft : StatefulTractogram
+        The stateful tractogram to save
+    filename : string
+        Filename with valid extension
+
+    Returns
+    -------
+    output : bool
+        Did the saving work properly
+    """
+
+    _, extension = os.path.splitext(filename)
+    if extension not in ['.trk', '.tck', '.vtk', '.fib', '.dpy']:
+        TypeError('Output filename is not one of the supported format')
+
+    if bbox_valid_check and not sft.is_bbox_in_vox_valid():
+        raise ValueError('Bounding box is not valid in voxel space, cannot ' +
+                         'save a valid file if some coordinates are invalid')
+
+    old_space = deepcopy(sft.get_current_space())
+    old_shift = deepcopy(sft.get_current_shift())
+
+    sft.to_rasmm()
+    sft.to_center()
+
+    timer = time.time()
+    if extension in ['.trk', '.tck']:
+        tractogram_type = detect_format(filename)
+        header = create_tractogram_header(tractogram_type,
+                                          *sft.get_space_attribute())
+        new_tractogram = Tractogram(sft.get_streamlines(),
```

Comment:
Can't you reuse `self._tractogram`? Why do you need to create a new one?

Reply:
I could, I wanted to minimize the use (and visibility) of _tractogram outside of the class.
This function is not part of the class, so it would not be a self and I preferred not to access the private tractogram variable.

Also I decided not to include load/save in the class so they can replace the older function (that have the same name) elsewhere. I think it would be a smoother transition.

Reply:
tractogram.affine_to_rasmm can be set properly, but then if someone call .to_world(), tractogram.affine_to_rasmm becomes identity (which would be problematic for the stateful_tractogram). And since I need the voxel size for some transformations anyway, I needed to store variables in the stateful_tractogram.

For technical clarity, I think that stateful_tractogram should store its own "spatial_attribute" and be free of the possibility of someone calling .to_word() that would modifying it.